velox/dwio/dwrf/reader/SelectiveColumnReader.cpp

```diff
@@ -3861,6 +3862,294 @@ void SelectiveStringDictionaryColumnReader::ensureInitialized() {
+  VELOX_CHECK(!positionsProvider.hasNext());
+}
+
+template <bool dense>
+void SelectiveTimestampColumnReader::readHelper(RowSet rows) {
+  vector_size_t numRows = rows.back() + 1;
+  ExtractToReader extractValues(this);
+  common::AlwaysTrue filter;
+  auto secondsV1 = dynamic_cast<RleDecoderV1<true>*>(seconds_.get());
+  VELOX_CHECK(secondsV1, "Only RLEv1 is supported");
+  if (nullsInReadRange_) {
+    secondsV1->readWithVisitor<true>(
+        nullsInReadRange_->as<uint64_t>(),
+        DirectRleColumnVisitor<
+            int64_t,
+            common::AlwaysTrue,
+            decltype(extractValues),
+            dense>(filter, this, rows, extractValues));
+  } else {
+    secondsV1->readWithVisitor<false>(
+        nullptr,
+        DirectRleColumnVisitor<
+            int64_t,
+            common::AlwaysTrue,
+            decltype(extractValues),
+            dense>(filter, this, rows, extractValues));
+  }
+
+  // Save the seconds into their own buffer before reading nanos into
+  // 'values_'
+  ensureCapacity<uint64_t>(secondsValues_, numValues_, &memoryPool_);
+  secondsValues_->setSize(numValues_ * sizeof(int64_t));
+  memcpy(
+      secondsValues_->asMutable<char>(),
+      rawValues_,
+      numValues_ * sizeof(int64_t));
+
+  // We read the nanos into 'values_' starting at index 0.
+  numValues_ = 0;
+  auto nanosV1 = dynamic_cast<RleDecoderV1<false>*>(nano_.get());
+  VELOX_CHECK(nanosV1, "Only RLEv1 is supported");
+  if (nullsInReadRange_) {
+    nanosV1->readWithVisitor<true>(
+        nullsInReadRange_->as<uint64_t>(),
+        DirectRleColumnVisitor<
+            int64_t,
+            common::AlwaysTrue,
+            decltype(extractValues),
+            dense>(filter, this, rows, extractValues));
+  } else {
+    nanosV1->readWithVisitor<false>(
+        nullptr,
+        DirectRleColumnVisitor<
+            int64_t,
+            common::AlwaysTrue,
+            decltype(extractValues),
+            dense>(filter, this, rows, extractValues));
+  }
+  readOffset_ += numRows;
+}
+
+void SelectiveTimestampColumnReader::read(
+    vector_size_t offset,
+    RowSet rows,
+    const uint64_t* incomingNulls) {
+  prepareRead<int64_t>(offset, rows, incomingNulls);
+  VELOX_CHECK(!scanSpec_->filter());
+  bool isDense = rows.back() == rows.size() - 1;
+  if (isDense) {
+    readHelper<true>(rows);
+  } else {
+    readHelper<false>(rows);
+  }
+}
+
+void SelectiveTimestampColumnReader::getValues(RowSet rows, VectorPtr* result) {
+  // We merge the seconds and nanos into 'values_'
+  auto tsValues = AlignedBuffer::allocate<Timestamp>(numValues_, &memoryPool_);
+  auto rawTs = tsValues->asMutable<Timestamp>();
+  auto secondsData = secondsValues_->as<int64_t>();
+  auto nanosData = values_->as<uint64_t>();
+  auto rawNulls = nullsInReadRange_
+      ? (returnReaderNulls_ ? nullsInReadRange_->as<uint64_t>()
+                            : rawResultNulls_)
+      : nullptr;
+  for (auto i = 0; i < numValues_; i++) {
```

Comment:
this is duplicate code and usually hard to reason about. Consider extracting it and share it across ColumnReader and SelectiveColumnReader